File: algorithms/sorting/bubble-sort/bubble-sort.mjs
/**
 * Return a sorted copy of the input using optimized Bubble Sort.
 * The original array is never mutated.
 */
export function bubbleSort(numbers) {
  const array = [...numbers];

  for (let pass = 0; pass < array.length - 1; pass += 1) {
    let swapped = false;

    for (let index = 0; index < array.length - pass - 1; index += 1) {
      if (array[index] > array[index + 1]) {
        [array[index], array[index + 1]] = [array[index + 1], array[index]];
        swapped = true;
      }
    }

    if (!swapped) break;
  }

  return array;
}

File: algorithms/sorting/bubble-sort/bubble_sort.cpp
#include <vector>

/** Return a sorted copy of the input without mutating it. */
std::vector<int> bubbleSort(const std::vector<int>& numbers) {
    std::vector<int> array = numbers;

    for (std::size_t pass = 0; pass + 1 < array.size(); ++pass) {
        bool swapped = false;

        for (std::size_t index = 0; index + pass + 1 < array.size(); ++index) {
            if (array[index] > array[index + 1]) {
                const int temporary = array[index];
                array[index] = array[index + 1];
                array[index + 1] = temporary;
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }
    }

    return array;
}
